pkg/terminal/p400/rabbit_service.go

```diff
@@ -0,0 +1,114 @@
+package p400
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// RabbitServicePayload represents the JSON shape of all request bodies when calling Rabbit Service
+type RabbitServicePayload struct {
+	ID            int         `json:"id"`
+	Service       string      `json:"service"`
+	Method        string      `json:"method"`
+	Content       string      `json:"content"`
+	SessionToken  string      `json:"session_token"`
+	RequestType   string      `json:"request_type"`
+	VersionInfo   versionInfo `json:"version_info"`
+	ParentTraceID string      `json:"parent_trace_id"`
+	DeviceInfo    DeviceInfo  `json:"device_info"`
+}
+
+// RabbitServiceResponse is the response body from the Rabbit Service call
+type RabbitServiceResponse struct {
+	Content string `json:"content"`
+}
+
+type versionInfo struct {
+	ClientType    string `json:"client_type"`
+	ClientVersion string `json:"client_version"`
+}
+
+func encodeRPCContent(rpcContent interface{}) (string, error) {
+	jsonContent, err := json.Marshal(rpcContent)
+
+	if err != nil {
+		return "", err
+	}
+
+	contentBytes := []byte(string(jsonContent))
+	encodedContent := base64.StdEncoding.EncodeToString(contentBytes)
+
+	return encodedContent, nil
+}
+
+// CallRabbitService takes a TerminalSessionContext and method information and calls that Rabbit Service RPC method
+func CallRabbitService(tsCtx TerminalSessionContext, method string, methodContent interface{}, methodResponse interface{}, parentTraceID string) error {
+	encodedMethodContent, err := encodeRPCContent(methodContent)
+
+	var result RabbitServiceResponse
+
+	if err != nil {
+		return err
+	}
+
+	payload := CreateRabbitServicePayload(method, encodedMethodContent, parentTraceID, tsCtx)
+	formattedIP := strings.Join(strings.Split(tsCtx.IPAddress, "."), "-")
+
+	rabbitServiceURL := fmt.Sprintf(readerURL, formattedIP)
+	res, err := http.Post(rabbitServiceURL, "application/json", &payload)
+
+	if err != nil {
+		return err
+	}
+
+	json.NewDecoder(res.Body).Decode(&result)
```

Comment:
possibly silly question, is there a difference between this and `json.Unmarshal(res.Body, &result)`?

Reply:
Great question. The response body is a stream rather than a string / string byte array so it was more convenient to pass it to a decoder instead. Is it more idiomatic to use `Unmarshal` after converting the stream?

Reply:
I read more into this and it seems that it's preferable just to pass it to a `Decoder`